src/sql-server-util/src/cdc.rs

```diff
@@ -310,60 +356,114 @@ pub enum CdcError {
+    pub fn try_from_bytes(bytes: &[u8]) -> Result<Self, String> {
+        if bytes.len() != Self::SIZE {
+            return Err(format!("incorrect length, expected 10 got {}", bytes.len()));
+        }
 
-    /// Returns `self` as a [`StructuredLsn`].
-    pub fn as_structured(&self) -> StructuredLsn {
-        let vlf_id: [u8; 4] = self.0[0..4].try_into().expect("known good length");
-        let block_id: [u8; 4] = self.0[4..8].try_into().expect("known good length");
-        let record_id: [u8; 2] = self.0[8..].try_into().expect("known good length");
+        let vlf_id: [u8; 4] = bytes[0..4].try_into().expect("known good length");
+        let block_id: [u8; 4] = bytes[4..8].try_into().expect("known good length");
+        let record_id: [u8; 2] = bytes[8..].try_into().expect("known good length");
 
-        StructuredLsn {
+        Ok(Lsn {
             vlf_id: u32::from_be_bytes(vlf_id),
             block_id: u32::from_be_bytes(block_id),
             record_id: u16::from_be_bytes(record_id),
-        }
+        })
+    }
+
+    /// Return the underlying byte slice for this [`Lsn`].
+    pub fn as_bytes(&self) -> [u8; 10] {
+        let mut raw: [u8; Self::SIZE] = [0; 10];
+
+        raw[0..4].copy_from_slice(&self.vlf_id.to_be_bytes());
+        raw[4..8].copy_from_slice(&self.block_id.to_be_bytes());
+        raw[8..].copy_from_slice(&self.record_id.to_be_bytes());
+
+        raw
     }
-}
 
-impl Ord for Lsn {
-    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        self.as_structured().cmp(&other.as_structured())
+    /// Increment this [`Lsn`].
+    ///
+    /// The returned [`Lsn`] may not exist upstream yet, but it's guaranteed to
+    /// sort greater than `self`.
+    pub fn increment(self) -> Lsn {
+        let overflow_record_id = match self.record_id.checked_add(1) {
+            Some(new_record_id) => {
+                return Lsn {
+                    vlf_id: self.vlf_id,
+                    block_id: self.block_id,
+                    record_id: new_record_id,
+                };
+            }
+            None => 0,
+        };
+        let overflow_block_id = match self.block_id.checked_add(1) {
+            Some(new_block_id) => {
+                return Lsn {
+                    vlf_id: self.vlf_id,
+                    block_id: new_block_id,
+                    record_id: overflow_record_id,
+                };
+            }
+            None => 0,
+        };
+        match self.vlf_id.checked_add(1) {
+            Some(new_vlf_id) => Lsn {
+                vlf_id: new_vlf_id,
+                block_id: overflow_block_id,
+                record_id: overflow_record_id,
+            },
+            None => panic!("overflowed Lsn, {self:?}"),
+        }
     }
 }
 
-impl PartialOrd for Lsn {
-    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
-        Some(self.cmp(other))
+impl fmt::Display for Lsn {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Comment:
Is there a more concise format than the debug view? This will be used to log frontiers and it's going to be annoying if it gets printed in such detail. In some online resources I've seen the format `format!({}:{}:{}", self.vlf_id, self.block_id, self.record_id)` which seems nice and compact. https://www.sqlshack.com/go-lsn-in-sql-server/

Reply:
Yeah the debug format was just easy, updated to your suggestion!